### PR TITLE
Revamp Folder Tree view (left pane) and the Main view (right pane)

### PR DESCRIPTION
**Pull request by hbjORbj**
Linear: https://linear.app/intuita/issue/INT-798/display-cases-under-folder-and-hide-jobs-under-case-allow-users-to
Miro: https://miro.com/app/board/o9J_kiRkEQ4=/?moveToWidget=3458764551403858729&cot=10
Screenshot 2023-04-14 at 13 47 25

### 1. View by codemod
- The view stays the same. The added functionality is that user will see a scrollable main view of file changes (rather than 1 file change) on pressing a case.
```
Upgrade Next to v13
   --- Case
      --- Job 1
      --- Job 2
      ...
```

### 2. View by folder
- The view changed. We do not show individual jobs in this view. Each folder contains case(s) that bring change to that folder. Also, same to the view by codemod, user will see a scrollable main view of file changes (rather than 1 file change) on pressing a case.
```
Upgrade Next to v13
   --- Root
      --- packages
         --- Case 1
         --- Case 2
      --- app
      ...
```

### CODE REVIEW

**Comment by hbjORbj** [comment]
### Done for folder breakdown view
- Folder breakdown structure is pretty solid, tested multiple times to make sure that no duplicate items or id bugs exist
- Applying / Discarding “by case”
- Diff view “by case”

### TODOs for folder breakdown view
- Creating issue / PR “by case”
- Applying / Discarding / Creating issue / PR “by folder”
- Make cases visible in all folders in which they exist

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.css [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+.collapsable__arrow {
+	margin-right: 0.5rem;
+	width: 18px;
+	height: 18px;
+	transition: transform 0.2s ease-in-out;
+}
+
+.collapsable__arrow--open {
+	transform: rotate(180deg);
+}
```
Comment (hbjORbj) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.css:24]: Not needed if we use vscode native icon.

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import { forwardRef, useImperativeHandle, useState } from 'react';
+import { ReactComponent as ArrowDownIcon } from '../../assets/arrow-down.svg';
```
Comment (hbjORbj) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:2]: Not needed

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			<div className={`collapsable ${className ?? ''} `}>
+				<div
+					className={`collapsable__header ${headerClassName ?? ''} `}
+					onClick={() => setExpanded(!expanded)}
+				>
+					<ArrowDownIcon
+						className={`collapsable__arrow ${
+							expanded ? 'collapsable__arrow--open' : ''
+						}`}
+					/>
```
Comment (hbjORbj) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:50]:
```suggestion
					<span
						className={cn(
							'collapsable__arrow',
							'codicon',
							expanded
								? 'codicon-arrow-bottom'
								: 'codicon-arrow-right',
						)}
					/>
```

File: intuita-webview/src/jobDiffView/App.tsx [hunk 4]
```diff
@@ -12,45 +12,54 @@ const getViewComponent = (view: View) => {
...
-					});
+			if (
+				message.kind === 'webview.diffView.updateDiffViewProps' &&
+				view?.viewId === 'jobDiffView'
+			) {
+				const jobHash = message?.data?.jobHash;
+				if (!jobHash) {
+					return;
+				}
+				const viewData = [...view?.viewProps?.data];
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/App.tsx:37]: `view.viewProps.data.slice() ?? []`
I think that `?` should not be required at the line

File: intuita-webview/src/jobDiffView/App.tsx [hunk 5]
```diff
@@ -12,45 +12,54 @@ const getViewComponent = (view: View) => {
...
+				const index = viewData.findIndex(
+					(el) => el.jobHash === jobHash,
+				);
+				if (index === -1) {
+					return;
 				}
+				viewData[index] = message.data;
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/App.tsx:44]: You shouldn't use mutable operations at all, better to clone the array and splice it later on.

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
+import { forwardRef, useImperativeHandle, useState } from 'react';
+import { ReactComponent as ArrowDownIcon } from '../../assets/arrow-down.svg';
+import './Collapsable.css';
+
+type CollapsableProps = {
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:5]: `Readonly<...>`

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	defaultExpanded: boolean;
+	headerComponent: React.ReactNode;
+	headerClassName?: string;
+	children: React.ReactNode;
+	contentClassName?: string;
+	className?: string;
+};
+
+export type CollapsableRefMethods = {
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:14]: `Readonly<...>`

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+			headerClassName,
+			contentClassName,
+			className,
+			children,
+		},
+		ref,
+	) => {
+		const [expanded, setExpanded] = useState(defaultCollapsed);
+
+		useImperativeHandle(ref, () => ({
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:34]: can you explain why `ref` is needed?

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+			expanded,
+			collapse: () => setExpanded(false),
+			expand: () => setExpanded(true),
+		}));
+
+		return (
+			<div className={`collapsable ${className ?? ''} `}>
+				<div
+					className={`collapsable__header ${headerClassName ?? ''} `}
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:43]: can you use `classnames` lib here?

**Review by grzpab — COMMENTED** [review]
I haven't reviewed all the code yet but almost ready.

File: intuita-webview/src/assets/arrow-down.svg
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/assets/arrow-down.svg:1]: @henoktsegayetoptal

File: intuita-webview/src/jobDiffView/App.tsx [hunk 10]
```diff
@@ -12,45 +12,54 @@ const getViewComponent = (view: View) => {
 	switch (view.viewId) {
 		case 'jobDiffView':
 			const { data } = view.viewProps;
-			return <JobDiffView {...data} />;
+			return data.map((el) => <JobDiffView key={el.jobHash} {...el} />);
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/App.tsx:15]: @henoktsegayetoptal 's part
Benny: solved it

File: intuita-webview/src/jobDiffView/hooks/useElementSize.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import { useState, useEffect, RefObject } from 'react';
+
+type Size = Readonly<{
+	width: number;
+	height: number;
+}>;
+export function useElementSize<T extends HTMLElement>(ref: RefObject<T>): Size {
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/hooks/useElementSize.ts]: @henoktsegayetoptal 's part
Benny: Hmm, can't find a similar function.

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+	children,
+}: CollapsableProps) => {
+	const [isOpen, setIsOpen] = useState(false);
+	return (
+		<div className={`collapsable ${className ?? ''} `}>
+			<div
+				className={`collapsable__header ${headerClassName ?? ''} `}
+				onClick={() => setIsOpen(!isOpen)}
+			>
+                <ArrowDownIcon className={`collapsable__arrow ${isOpen ? 'collapsable__arrow--open' : ''}`} />
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx]: @henoktsegayetoptal 's part
Benny: solved it

File: intuita-webview/src/jobDiffView/App.tsx [hunk 13]
```diff
@@ -12,45 +12,54 @@ const getViewComponent = (view: View) => {
...
-							data: {
-								...data,
-							},
-						},
-					});
+			if (
+				message.kind === 'webview.diffView.updateDiffViewProps' &&
+				view?.viewId === 'jobDiffView'
+			) {
+				const jobHash = message?.data?.jobHash;
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/App.tsx:33]: @henoktsegayetoptal 's part
Benny: solved it

File: intuita-webview/src/jobDiffView/App.tsx [hunk 14]
```diff
@@ -12,45 +12,54 @@ const getViewComponent = (view: View) => {
...
+				if (!jobHash) {
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/App.tsx:34]: @henoktsegayetoptal 's part
Benny: solved it

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx (hunk 6 above)
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:5]: @henoktsegayetoptal 's part
Benny: addressed it

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx (hunk 7 above)
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:14]: @henoktsegayetoptal 's part
Benny: addressed it

File: intuita-webview/src/jobDiffView/App.tsx (hunk 4 above)
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/App.tsx:37]: @henoktsegayetoptal 's part
Benny: addressed it

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx (hunk 8 above)
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:34]: @henoktsegayetoptal

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx (hunk 9 above)
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:43]: @henoktsegayetoptal 's part
Benny: addressed it

File: intuita-webview/src/jobDiffView/DiffViewer/Container.tsx [hunk 15]
```diff
@@ -1,60 +1,98 @@
...
+		const handleChange = (e: Event | React.FormEvent<HTMLElement>) => {
+			const value = (e.target as HTMLSelectElement).value as
+				| 'inline'
+				| 'side-by-side';
+			onViewTypeChange(value);
+		};
+		return (
+			<div
+				className="flex  flex-wrap w-full container flex-col"
+				ref={ref}
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/Container.tsx:37]: @henoktsegayetoptal

File: intuita-webview/src/jobDiffView/DiffViewer/Diff.tsx [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+				wrapLines={true}
+				language="javascript"
+			>
+				{value}
+			</SyntaxHighlighter>
+		);
+	};
+
+	return (
+		<div className="w-full" ref={containerRef}>
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/Diff.tsx:38]: @henoktsegayetoptal

File: intuita-webview/src/jobDiffView/DiffViewer/index.tsx [hunk 17]
```diff
@@ -27,79 +25,49 @@ export const JobDiffView = ({
 			? 'inline'
 			: 'side-by-side';
 	});
+	const [viewed, setViewed] = useState(false);
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/index.tsx:28]: @henoktsegayetoptal 's part
Benny: addressed it

**Comment by hbjORbj** [comment]
@henoktsegayetoptal 

fyi, this commit addresses Greg's comments on your code. https://github.com/intuita-inc/intuita-vscode-extension/pull/251/commits/82ee77874272b8ded86774908b4d5abba47cc314

File: intuita-webview/.npmrc [hunk 18]
```diff
@@ -0,0 +1 @@
+auto-install-peers=true
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/.npmrc]: @henoktsegayetoptal 's part

Benny:
pnpm uses it: https://pnpm.io/npmrc
it looks like yarn uses `.yarnrc`.

File: intuita-webview/src/main/TreeView/index.tsx [hunk 19]
```diff
@@ -106,6 +106,7 @@ const TreeView = ({ node }: Props) => {
 				label={node.label ?? ''}
 				icon={icon}
 				depth={depth}
+				kind={node.kind ?? ''}
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply intuita-webview/src/main/TreeView/index.tsx:109]: No. I changed `TreeNode.kind` to required type.

File: src/components/webview/MainWebviewProvider.ts [hunk 20]
```diff
@@ -191,49 +180,96 @@ export class IntuitaProvider implements WebviewViewProvider {
 
 			// e.g., ['packages', 'app', 'src', 'index.tsx']
 			let directories = filePath.split('/').filter((item) => item !== '');
-			// e.g., ['/', 'packages', 'app', 'src', 'index.tsx']
-			directories = ['/', ...directories];
-			// e.g., 'index.tsx'
-			const fileName = directories[directories.length - 1];
+			// e.g., ['/', 'packages', 'app', 'src']
+			directories = ['/', ...directories.slice(0, -1)];
```
[the remark this answers is not in the document]
Comment (hbjORbj) [reply src/components/webview/MainWebviewProvider.ts:184]: This is how the feature works now. https://intuita-inc.slack.com/files/U0508J12DR8/F0530Q3CXEK/screen_recording_2023-04-16_at_16.28.59.mov

We no longer display the list of jobs under each file. So, we no longer need to iterate on the file itself. We only need folders.

File: intuita-webview/src/jobDiffView/App.tsx [hunk 21]
```diff
@@ -12,45 +12,64 @@ const getViewComponent = (view: View) => {
...
+			if (
+				message.kind === 'webview.diffView.updateDiffViewProps' &&
+				view.viewId === 'jobDiffView'
+			) {
+				const jobHash = message.data.jobHash ?? null;
+				if (jobHash === null) {
+					return;
 				}
+
+				const index = (view.viewProps.data.slice() ?? []).findIndex(
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/App.tsx:46]: ```view.viewProps.data.slice() ?? []``` needs no `?? []`

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+						className={cn(
+							'collapsable__arrow',
+							expanded && 'collapsable__arrow--open',
+						)}
+					/>
+					{headerComponent}
+				</div>
+				{expanded && (
+					<div
+						className={` collapsable_content ${
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:57]: you can use `cn` here as well

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+		const [expanded, setExpanded] = useState(defaultCollapsed);
+
+		useImperativeHandle(ref, () => ({
+			expanded,
+			collapse: () => setExpanded(false),
+			expand: () => setExpanded(true),
+		}));
+
+		return (
+			<div className={`collapsable ${className ?? ''} `}>
```
Comment (grzpab) [inline intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:42]: you can use `cn` here as well

File: src/components/webview/DiffWebviewPanel.ts [hunk 24]
```diff
@@ -123,6 +141,36 @@ export class DiffWebviewPanel extends IntuitaWebviewPanel {
 		};
 	}
 
+	public async getViewDataForCase(
+		caseHash: ElementHash,
+	): Promise<Readonly<JobDiffViewProps>[]> {
+		const jobHashes = Array.from(
+			this.__caseManager.getJobHashes([
+				caseHash,
+			] as unknown as CaseHash[]),
```
Comment (grzpab) [inline src/components/webview/DiffWebviewPanel.ts:150]: how can a job hash become a case hash? this looks like a serious issue to me

File: src/components/webview/MainWebviewProvider.ts [hunk 25]
```diff
@@ -664,4 +707,91 @@ export class IntuitaProvider implements WebviewViewProvider {
 		}
 		return mappedNode;
 	};
+
+	private __buildCaseByFolderTree = (
+		element: FileElement,
+		parentFolderPath: string,
+	): TreeNode | null => {
+		if (!element.children || element.children.length === 0) {
```
Comment (grzpab) [inline src/components/webview/MainWebviewProvider.ts:715]: now I am thinking we should ensure `element` children are always just an non-nullable array (let's do this change in another PR)

File: src/components/webview/DiffWebviewPanel.ts (hunk 24 above)
Comment (hbjORbj) [reply src/components/webview/DiffWebviewPanel.ts:150]: `caseHash` is of type `ElementHash`.

```jsx
	public async getViewDataForCase(
		caseHash: ElementHash,
	): Promise<Readonly<JobDiffViewProps>[]> {
		const jobHashes = Array.from(
			this.__caseManager.getJobHashes([
				caseHash,
			] as unknown as CaseHash[]),
		);

		if (jobHashes.length === 0) {
			return [];
		}
		const viewDataArray = await Promise.all(
			jobHashes.map((jobHash) => this.getViewDataForJob(jobHash)),
		);
		return viewDataArray.filter(isNeitherNullNorUndefined);
	}
```

File: intuita-webview/src/jobDiffView/App.tsx (hunk 21 above)
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/App.tsx:46]: addressed

File: intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx (hunk 23 above)
Comment (hbjORbj) [reply intuita-webview/src/jobDiffView/DiffViewer/Collapsable.tsx:42]: addressed

File: src/components/webview/DiffWebviewPanel.ts (hunk 24 above)
Comment (hbjORbj) [reply src/components/webview/DiffWebviewPanel.ts:150]: I separated the original function that receives an argument of type `ElementHash | JobHash[]` into 2 functions:

```jsx
	public async getViewDataForCase(
		caseHash: ElementHash,
	): Promise<Readonly<JobDiffViewProps>[]> {
		const jobHashes = Array.from(
			this.__caseManager.getJobHashes([
				caseHash,
			] as unknown as CaseHash[]),
		);

		if (jobHashes.length === 0) {
			return [];
		}
		const viewDataArray = await Promise.all(
			jobHashes.map((jobHash) => this.getViewDataForJob(jobHash)),
		);
		return viewDataArray.filter(isNeitherNullNorUndefined);
	}

	public async getViewDataForJobsArray(
		jobHashes: JobHash[],
	): Promise<Readonly<JobDiffViewProps>[]> {
		if (jobHashes.length === 0) {
			return [];
		}
		const viewDataArray = await Promise.all(
			jobHashes.map((jobHash) => this.getViewDataForJob(jobHash)),
		);
		return viewDataArray.filter(isNeitherNullNorUndefined);
	}
```

**Review by grzpab — APPROVED** [review]
